src/index/base.cpp

```diff
@@ -67,10 +69,16 @@ bool BaseIndex::Init()
     if (locator.IsNull()) {
         SetBestBlockIndex(nullptr);
     } else {
-        SetBestBlockIndex(m_chainstate->FindForkInGlobalIndex(locator));
+        // Setting the best block to the locator's top block. If it is not part of the
+        // best chain, we will rewind to the fork point during index sync
+        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator.vHave.front())};
```

Comment:
Just leaving as a note, but I think this allows the `if (!active_chain.Contains(block_to_test)) {` case to be hit. Previously `FindForkInGlobalIndex` would return a `CBlockIndex` only in the current chain and so `active_chain.Contains` would always be true.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r904094952

In commit "index: Use first block from locator instead of looking for fork point" (20bd2219894e53ab39d2b11edf9a7f3c7bda489b)

Yes, this seems to be the case. This is a good indication the original author or this code expecting was expecting to FindForkInGlobalIndex to just return a block based on the locator, not necessarily a block on the current chain. So new code is probably closer to intent of original code.

Reply:
Thank you your a angle

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r907572547

> Without this patch, is it possible that the old code goes back to genesis in this rare, kind of contrived scenario?:

Maybe I need to think about this more, but I don't think it can go back to genesis. It can just go back to the last common block before the reorg. Also, the problem which this patch fixes isn't going backwards in general, but going backwards without rewinding. The problem with the `FindForkInGlobalIndex` is that it goes backwards without making needed `Rewind()` call to update indexes

Reply:
From my reading, if `FindForkInGlobalIndex` doesn't find `pindex` in the chain or doesn't find `Tip()` as an ancestor of `pindex`, it returns genesis. But yeah a bit orthogonal to the issue at hand

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r907779757

> From my reading, if `FindForkInGlobalIndex` doesn't find `pindex` in the chain or doesn't find `Tip()` as an ancestor of `pindex`, it returns genesis. But yeah a bit orthogonal to the issue at hand

Wow, you are right. The [`FindForkInGlobalIndex`](https://github.com/bitcoin/bitcoin/blob/2111f32f2a6998531871e7792b5208992868ba7f/src/validation.cpp#L131-L149) behavior is much different than the [`FindFork`](https://github.com/bitcoin/bitcoin/blob/2111f32f2a6998531871e7792b5208992868ba7f/src/chain.cpp#L54-L63) behavior. I assumed `FindForkInGlobalIndex` would try to find the last common ancestor between the locator block and the chain like `FindFork` does, but it doesn't even try to do this. Instead it will literally only return the exact locator block, or the chain tip, or the genesis block. It doesn't make any sense to me why the function would be implemented this way, but I guess it's good that this PR removes one usage of it...